clis/teliod/src/main.rs

```diff
@@ -115,17 +152,188 @@ impl CommandListener {
+                        &client_config.auth_token,
+                    )
+                    .await?
+                }
+            }
+    }
+
+    let mut retries = 0;
+    loop {
+        let status = update_machine(&client_config).await?;
+        match status {
+            StatusCode::OK => break,
+            StatusCode::FORBIDDEN | StatusCode::UNAUTHORIZED | StatusCode::BAD_REQUEST => {
+                // exit daemon
+                return Err(TeliodError::UpdateMachineError(status));
+            }
+            StatusCode::NOT_FOUND => {
+                debug!("Unable to update. Registering machine ...");
+                // Retry machine update after registering. To make sure everything was successful
+                // If registering fails. Close the daemon
+                client_config.machine_identifier = register_machine(
+                    &client_config.hw_identifier,
+                    client_config.public_key,
+                    &client_config.auth_token,
+                )
+                .await?;
+            }
+            _ => {
+                // Retry with exp back-off
+                if retries < MAX_RETRIES {
+                    let backoff_time = BASE_DELAY_MS * 2_u64.pow(retries);
+                    let sleep_time = min(backoff_time, MAX_BACKOFF_TIME);
+                    debug!("Retrying after {} ms...", sleep_time);
+                    tokio::time::sleep(Duration::from_millis(sleep_time)).await;
+                    retries += 1;
+                } else {
+                    // If max retries exceeded, exit daemon
+                    return Err(TeliodError::UpdateMachineTimeout);
+                }
+            }
+        }
+    }
+
+    config.private_key = Some(client_config.private_key.clone());
+    config.hw_identifier = Some(client_config.hw_identifier.clone());
+    config.machine_identifier = Some(client_config.machine_identifier.clone());
+
+    let mut file = File::create(config_path)?;
+    serde_json::to_writer(&mut file, &config)?;
+    Ok(client_config)
+}
+
 // From async context Telio needs to be run in separate task
-fn telio_task() {
-    let _telio = Device::new(
+fn telio_task(
+    client_config: ClientConfig,
+    mut rx_channel: mpsc::Receiver<TelioTaskCmd>,
+    tx_channel: mpsc::Sender<TelioTaskCmd>,
+) -> Result<(), TeliodError> {
+    debug!("Initializing telio device");
+    let mut telio = Device::new(
         Features::default(),
         // TODO: replace this with some real event handling
         move |event| info!("Incoming event: {:?}", event),
         None,
-    );
+    )?;
+
+    let config_ptr = std::sync::Arc::new(client_config);
+    // TODO: This is temporary to be removed later on when we have proper integration
+    // tests with core API. This is to not look for tokens in a test environment
+    // right now as the values are dummy and program will not run as it expects
+    // real tokens.
+    if !config_ptr.auth_token.eq("") {
```

Comment:
Would be nice to also get (or override) this from env variable `NORD_TOKEN` if present, like tcli does.

Reply:
Done!